datapath: Switch to LPM policy map

jrajahalme opened the pull request:
Change to use LPM trie map for endpoint policy, as this allows prefix based wildcarding of protocol and port, when placed at the end of the map key. This helps eliminate two policy map lookups, which should help offset the performance penalty of a more complex map lookup.

This involves both changing the map type, and extending the policymap key with the prefix length field that is customary to all LPM maps. At the same time we need to reorganise the key fields so that the fields we plan to wildcard are at the end of the key and in the order of the port being the last field and protocol before it, as it can be wildcarded also when the protocol is not wildcarded, but the protocol can be wildcarded only if the port is also (completely) wildcarded.

With this new LPM policy map we still need a second lookup with an explicitly wildcarded security ID, as it can be wildcarded also when protocol and/or port are being matched.

As Cilium agent can re-populate policy maps from scratch there is no need for any specific upgrade/downgrade support for the map type and key layout change. Old programs keep using the old policy maps they were started with, while new programs for all endpoints upon the start of the upgraded Cilium agent will the new maps when they are loaded.

This change is a prerequisite for port range support in network policy.

thorn3r commented on cilium/cmd/bpf_policy_get.go:
```diff
@@ -141,21 +142,30 @@ func formatMap(w io.Writer, statsMap []policymap.PolicyEntryDump) {
+			} else if prefix == 8 {
+				port = fmt.Sprintf("*/%s", proto.String())
+			} else if prefix > 8 {
+				port = fmt.Sprintf("0x%x/%d/%s", dport, prefix-8, proto.String())
```
do we care here about values > 24? i.e. are they valid, or are we using prefixlen to bound the value?

gentoo-root replied:
As far as I understood, 24 = 16+8, i.e. it's the maximum length for the port and protocol fields.

thorn3r replied:
thats how it seems to me as well. i guess my question could be rephrased a bit clearer as: 
is it possible to encounter a value > 24 that can cause some unexpected behavior here?

jrajahalme replied:
Added safety checks, thanks!

thorn3r reviewed (commented):
Agent bits LGTM, but i see some CI failures that look maybe relevant to these changes:
```
Tests upgrade and downgrade from a Cilium stable image to master:
Key-size mismatch for BPF map" file-path=/sys/fs/bpf/tc/globals/cilium_policy_00835 new=8 old=12 subsys=bpf
```

gentoo-root commented on bpf/lib/maps.h:
```diff
@@ -75,15 +75,21 @@ struct {
 	__uint(pinning, LIBBPF_PIN_BY_NAME);
 	__uint(max_entries, POLICY_MAP_SIZE);
-	__uint(map_flags, CONDITIONAL_PREALLOC);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
```
Why is this flag needed unconditionally now?

jrajahalme replied:
LPM maps can not have preallocated storage.

gentoo-root commented on bpf/lib/common.h:
```diff
@@ -318,17 +318,20 @@ struct remote_endpoint_info {
 };
 
 struct policy_key {
+	struct bpf_lpm_trie_key lpm_key;
```
It doesn't seem to be used when HAVE_LPM_TRIE_MAP_TYPE is 0. Put it under an `#ifdef`? Will the lookups be faster with a smaller key?

jrajahalme replied:
We now assume LPM support, so this comment does not apply any more.

gentoo-root commented on bpf/lib/policy.h:
```diff
@@ -234,11 +293,73 @@ __policy_can_access(const void *map, struct __ctx_buff *ctx, __u32 local_id,
+	/* This is a full L3/L4 match if port is not wildcarded,
+	 * need to check for L4-only policy first if it is.
+	 */
+	if (likely(policy && !policy->wildcarded_port)) {
```
The check for `is_untracked_fragment` seems to be missing from this implementation.

jrajahalme replied:
It is not needed, there is a note about it:
```
	 * Untracked fragments always have zero ports in the tuple so they can
	 * only match entries that have fully wildcarded ports.
```

gentoo-root commented on bpf/lib/policy.h:
```diff
@@ -234,11 +293,73 @@ __policy_can_access(const void *map, struct __ctx_buff *ctx, __u32 local_id,
+		policy = l4policy;
+		goto policy_check_entry;
+	}
+#endif
```
This lookup procedure, that repeats in policy_sk_egress, is a good candidate to be separated into an inline function. Is it feasible to do it in this pull request?

jrajahalme replied:
`policy_sk_egress()` was recently removed, so now there is only one function doing the policy lookup.

gentoo-root commented on pkg/maps/policymap/policymap.go:
```diff
@@ -237,10 +270,14 @@ func (key *PolicyKey) GetKeyPtr() unsafe.Pointer { return unsafe.Pointer(key) }
+		if prefixLen == 16 {
+			return fmt.Sprintf("%s: %d %d/%d", trafficDirectionString, key.Identity, byteorder.NetworkToHost16(key.DestPort), key.Nexthdr)
+		} else {
+			return fmt.Sprintf("%s: %d %d/%d/%d", trafficDirectionString, key.Identity, key.Nexthdr, byteorder.NetworkToHost16(key.DestPort), prefixLen)
```
Sorry my ignorance, but why do `Nexthdr` and `byteorder.NetworkToHost16(key.DestPort)`come in different order depending on prefixLen?

jrajahalme replied:
changed them back to the original order.

gentoo-root commented on cilium/cmd/helpers.go:
```diff
@@ -344,15 +344,15 @@ func updatePolicyKey(pa *PolicyUpdateArgs, add bool) {
+				err = policyMap.Deny(0, pa.label, pa.port, u8p, pa.trafficDirection)
 			} else {
-				err = policyMap.Allow(pa.label, pa.port, u8p, pa.trafficDirection, authType, proxyPort)
+				err = policyMap.Allow(0, pa.label, pa.port, u8p, pa.trafficDirection, authType, proxyPort)
```
Why is it 0, isn't it supposed to be 24 to include port and protocol?

jrajahalme replied:
This was a half-baked attempt to model an API where the prefix length is exposed to allow port ranges to be supported later. Removed this so that this PR is much cleaner!

gentoo-root reviewed (changes requested):
There are also a couple of questions, sorry in advance if they are too stupid, I'm new to this code area.

jrajahalme committed:
policymap: Keep key and value always in network byteorder

Reduce confusion by always keeping the policymap key and value in network
byteorder.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
datapath: Switch to LPM policy map

Change to use LPM trie map for endpoint policy, as this allows prefix
based wildcarding of protocol and port, when placed at the end of the map
key. This helps eliminate two policy map lookups, which should help
offset the performance penalty of a more complex map lookup.

This change is a prerequisite for port range support in network policy.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

gentoo-root commented on pkg/maps/policymap/policymap.go:
```diff
@@ -237,89 +254,93 @@ func (key *PolicyKey) GetKeyPtr() unsafe.Pointer { return unsafe.Pointer(key) }
-		return fmt.Sprintf("%s: %d %d/%d", trafficDirectionString, key.Identity, byteorder.NetworkToHost16(key.DestPort), key.Nexthdr)
+	if prefixLen > 0 {
+		dport := key.GetDestPort()
+		if prefixLen == 16 {
```
Am I missing something, or prefixLen == 16 means half of the dport field is wildcarded? struct policy_key:

```
__u8		protocol; /* can be wildcarded if 'dport' is fully wildcarded */
__u16		dport; /* can be wildcarded with CIDR-like prefix */
```

protocol goes first, so when both protocol and dport are non-wildcarded, prefixLen is 24, so I don't seem to understand why the check is for 16 here.

jrajahalme replied:
You are right, this needs to be 24 for the "full prefix" instead of 16, fixed :-)

jrajahalme committed:
datapath: Use inline function to keep policy and l4policy checks separate

bpf unit tests failed with a null pointer access verifier error (R1
invalid mem access 'inv'). Apparently the assignment of 'l4policy' to
'policy' was confusing, so separate the code path for 'l4policy' to keep
verifier happy.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

squeed reviewed (approved):
Looks good for my files.

thorn3r reviewed (approved):
LGTM :+1:

jrajahalme committed:
test: Ignore policy map mismatch warnings

Do not fail tests for policy map key or value size mismatches, as in case
of policy maps the maps are regenerated from the agent without any help
from the datapath.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

aditighag reviewed (changes requested):
>Change to use LPM trie map for endpoint policy, as this allows prefix
based wildcarding of protocol and port, when placed at the end of the map
key.

This commit is missing a lot of context. The commit shuffles some of the fields in policy key around. So how do we ensure that the datapath is in sync with the control plane state? I would've expected to see some map versioning code where the agent would populate the v2 policy map with the new key layout, and then atomically swap the old map. If the commit takes a different approach, then it needs to be explicitly called out in the commit description and code. 

Also, please mention in the code why the port and protocol fields need to be at the end of the key struct.

aditighag commented on bpf/lib/common.h:
```diff
@@ -316,18 +316,44 @@ struct remote_endpoint_info {
+ * order, as we want to be able to wildcard those fields in a specific pattern:
+ * 'protocol' can only be wildcarded if dport is also fully wildcarded.
+ * 'protocol' is never partially wildcarded, so it is either fully wildcarded or
+ * not wildcarded at all. 'dport' can be partially wildcarded, but only when
```
What does partially wildcarded mean? Does it indicate a port range?

jrajahalme replied:
This PR is a prerequisite for implementing port ranges, yes.

aditighag commented on pkg/endpoint/bpf.go:
```diff
@@ -1089,12 +1089,8 @@ func (e *Endpoint) updatePolicyMapPressureMetric() {
-		Nexthdr:          keyToDelete.Nexthdr,
-		TrafficDirection: keyToDelete.TrafficDirection,
-	}
+	policymapKey := policymap.NewKey(keyToDelete.Identity, keyToDelete.DestPort,
```
nit: Please revert the unintended alignment changes. Same for lines below.

jrajahalme replied:
This change is more than an alignment change, it is a change from a struct to a function call. While member per line makes sense for a struct initialization as it mirrors the struct definition, there is no such correspondence for function call. We typically break up function call paramaters to different lines if lines get long.

jrajahalme replied:
So the change here (and below) was intended.

jrajahalme replied:
same here.

aditighag commented on pkg/maps/policymap/policymap.go:
```diff
@@ -237,89 +254,93 @@ func (key *PolicyKey) GetKeyPtr() unsafe.Pointer { return unsafe.Pointer(key) }
+	// This will have to be exposed to the caller when port ranges are supported.
+	prefixLen := GetStaticPrefixBits()
+	if proto != 0 {
+		prefixLen += 8
```
Can these be derived using `sizeof` instead of hard cording the sizes?

jrajahalme replied:
Done.

aditighag commented on cilium/cmd/bpf_policy_get.go:
```diff
@@ -141,25 +141,40 @@ func formatMap(w io.Writer, statsMap []policymap.PolicyEntryDump) {
+						port = fmt.Sprintf("%d/%s", dport, protoStr)
+					}
+				}
+			}
```
I believe this is the core change of the PR that lays the foundation for protocol/port wildcarding. Can we have a switch case here to decide which field(s) are wildcarded, and annotate the cases with "protocol wildcarded (prefix len == 8), etc"? 
Also, can we not use hardcoded values here (how about using `sizeof`)?

jrajahalme replied:
Refactored this to be done once in policymap package, in the form suggested here (switch case with comments, using consts defined with Sizeof instead of hard coding the field sizes.

aditighag commented on bpf/lib/common.h:
```diff
@@ -316,18 +316,44 @@ struct remote_endpoint_info {
+ * 'protocol' can only be wildcarded if dport is also fully wildcarded.
+ * 'protocol' is never partially wildcarded, so it is either fully wildcarded or
+ * not wildcarded at all. 'dport' can be partially wildcarded, but only when
+ * 'protocol' is fully specified. This follows the logic that the destination
+ * port is a property of a transport protocol and can not be specified without
+ * also specifying the protocol.
```
Can we have unit tests to validate this? I suppose verification can be done based on `wildcard_protocol` and `wildcard_port` fields in the policy map entries.

jrajahalme replied:
This is part of the contract between the agent and the datapath, so the (unit) testing needs to be on the agent side. Added unit tests to `pkg/policymap`, also added corresponding checks to tests in `pkg/policy`.

aditighag reviewed (changes requested):
Thanks for adding detailed comments around the port and protocol fields.
Re: map layout change: I suppose the agent populates a new map based on the revised key/value entries, and then swaps out the old map? While this PR doesn't change the logic explicitly, it does change the map type in addition to map key layout. 
(The context needs to go in the PR/commit description. )

jrajahalme committed:
policymap: Add IsDeny() to PolicyEntry to hide Flags

Use PolicyEntry level accessor for IsDeny() so that the callers do not need to
know about the flags explicitly.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policymap: Hide and simplify flags

Do not export flags definition, simplify via using custom type instead of
depending on conversions to uint8.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policymap: Add unit tests on map key/value invariants

Add policymap unit testing to validate these policymap key & value invariants:
- protocol/nexthdr can only be wildcarded if destination port is wildcarded
- value has wildcardNexthdr flag set if and only if key wildcards Nexthdr
- value has wildcardDestPort flag set if and only if key wildcards DestPort
- deny entries have no redirection nor auth type

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policymap: Use string for protocol in policy key

Use string for protocol in policymap key conversion to disambiguate port
and protocol numbers, and always include it in the string.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

aditighag commented on pkg/maps/policymap/policymap.go:
```diff
@@ -236,90 +251,121 @@ func (p PolicyEntriesDump) Less(i, j int) bool {
-	trafficDirectionString := (trafficdirection.TrafficDirection)(key.TrafficDirection).String()
-	if key.DestPort != 0 {
-		return fmt.Sprintf("%s: %d %d/%d", trafficDirectionString, key.Identity, byteorder.NetworkToHost16(key.DestPort), key.Nexthdr)
+	switch {
+	case prefixLen == 0, prefixLen == NexthdrBits:
+		// Protocol wildcarded or specified, wildcarded port
+		return protoStr
+	case prefixLen > NexthdrBits && prefixLen < FullPrefixBits:
+		// Protocol specified, partially wildcarded port
+		return fmt.Sprintf("0x%x/%d/%s", dport, prefixLen-NexthdrBits, protoStr)
+	case prefixLen == FullPrefixBits:
+		// Both protocol and port specified, nothing wildcarded
+		return fmt.Sprintf("%d/%s", dport, protoStr)
+	default:
+		// Invalid prefix length
+		return fmt.Sprintf("<INVALID PREFIX LENGTH: %d>", prefixLen)
```
Looks much better, and easy to grok. Thanks!

aditighag reviewed (approved):
Policy engine changes look good for my code owners. Thanks!
